opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java

```diff
@@ -389,7 +389,6 @@ private String getFirstRevision(String fullpath) throws IOException {
                 ensureCommand(CMD_PROPERTY_KEY, CMD_FALLBACK),
                 "rev-list",
                 "--reverse",
-                "--max-count=1",
```

Comment:
No big problem here but some food for thought:

* `Executor.exec()` and `Executor.exec(boolean)` actually spool the entire output of a program into memory. Probably that's not a burdensome amount with this removal of `--max-count`, given this is a `rev-list` of a single file.
* When the spooled bytes are read later at line 402, they are read with the default Java charset, which is locale dependent on the OS installation. For hashes of a `rev-list`, that is perhaps always OK — but I'm just remarking since elsewhere in `GitRepository`, specific effort is made to use a correct charset (see `newLogReader()`).
* Technically when read later at line 402, a `BufferedReader` is overkill, given that the bytes are already fully spooled in-memory.
* Indeed all the output of a `Process` (via `exec()`) must be read or else a deadlock could happen as one awaits the process exit value. The default spool does consume everything.

I wonder if this is an opportunity to create a `HeadHandler` implementation of `Executor.StreamHandler` that is line-oriented (e.g. characters out of bytes), that demands a specified encoding instead of allowing the default Java charset, and that reads and stores a maximum number of lines and then just consumes and discards bytes (without bothering with character conversion) to exhaust the stream and avoid allocating heap.

Reply:
Definitely good idea, I'd go for LineHandler (or similar) and then derive the HeadHandler out of it. For simplicity a lot of code could benefit from this.

Reply:
Thank you, @tulinkry .

Maybe the reverse though: a `LineHandler` might be an unbounded `HeadHandler`. The latter would need to interact low-level with the stream in order to be able to read and dump bytes when a line limit is reached.

Reply:
Thanks guys. It occurred to me that this might be a problem but pushed the resolution aside. Will look into it.

Reply:
The latest changes have the basic `HeadHandler` (sans encoding).